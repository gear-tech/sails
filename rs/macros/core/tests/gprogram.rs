use proc_macro2::TokenStream;
use quote::quote;
use sails_macros_core::__gprogram_internal as gprogram;

#[test]
fn generates_init_for_single_ctor() {
    let input = quote! {
        impl MyProgram {
            pub async fn new(p1: u32, p2: String) -> Self {
                Self { p1, p2 }
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_init_for_multiple_ctors() {
    let input = quote! {
        impl MyProgram {
            pub async fn new(p1: u32, p2: String) -> Self {
                Self { p1, p2 }
            }

            pub fn new2(p2: String, p1: u32) -> Self {
                Self { p1, p2 }
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_init_for_no_ctor() {
    let input = quote! {
        impl MyProgram {
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_for_single_service_with_non_empty_route() {
    let input = quote! {
        impl MyProgram {
            pub fn service(&self) -> MyService {
                MyService
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_for_multiple_services_with_non_empty_routes() {
    let input = quote! {
        impl MyProgram {
            #[route("svc1")]
            pub fn service1(&self) -> MyService {
                MyService
            }

            pub fn service2(&self) -> MyService {
                MyService
            }
        }
    };

    let result = gprogram(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_with_gprogram_attributes() {
    let args = quote!(
        handle_reply = my_handle_reply,
        handle_signal = my_handle_signal
    );
    let input = quote! {
        impl MyProgram {}
    };

    let result = gprogram(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn generates_handle_with_crate_path() {
    let args = quote!(crate = sails_rename,);
    let input = quote! {
        impl MyProgram {}
    };

    let result = gprogram(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
