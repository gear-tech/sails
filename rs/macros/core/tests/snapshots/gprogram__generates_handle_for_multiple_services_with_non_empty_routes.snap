---
source: rs/macros/core/tests/gprogram.rs
expression: result
---
impl MyProgram {
    fn __service1(&self) -> MyService {
        MyService
    }
    fn __service2(&self) -> MyService {
        MyService
    }
    pub fn create() -> Self {
        Self
    }
    #[route("svc1")]
    pub fn service1(
        &self,
    ) -> <MyService as sails_rs::gstd::services::Service>::Exposure {
        let service = self.__service1();
        let exposure = <MyService as sails_rs::gstd::services::Service>::expose(
            service,
            sails_rs::gstd::Syscall::message_id(),
            __ROUTE_SVC1.as_ref(),
        );
        exposure
    }
    pub fn service2(
        &self,
    ) -> <MyService as sails_rs::gstd::services::Service>::Exposure {
        let service = self.__service2();
        let exposure = <MyService as sails_rs::gstd::services::Service>::expose(
            service,
            sails_rs::gstd::Syscall::message_id(),
            __ROUTE_SERVICE2.as_ref(),
        );
        exposure
    }
}
const __ROUTE_SVC1: [u8; 5usize] = [16u8, 83u8, 118u8, 99u8, 49u8];
const __ROUTE_SERVICE2: [u8; 9usize] = [
    32u8, 83u8, 101u8, 114u8, 118u8, 105u8, 99u8, 101u8, 50u8,
];
impl sails_rs::meta::ProgramMeta for MyProgram {
    type ConstructorsMeta = meta_in_program::ConstructorsMeta;
    const SERVICES: &'static [(&'static str, sails_rs::meta::AnyServiceMetaFn)] = &[
        ("Svc1", sails_rs::meta::AnyServiceMeta::new::<MyService>),
        ("Service2", sails_rs::meta::AnyServiceMeta::new::<MyService>),
    ];
    const ASYNC: bool = <MyService as sails_rs::meta::ServiceMeta>::ASYNC
        || <MyService as sails_rs::meta::ServiceMeta>::ASYNC;
}
mod meta_in_program {
    use super::*;
    use sails_rs::gstd::InvocationIo;
    #[derive(sails_rs::Decode, sails_rs::TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __CreateParams {}
    impl InvocationIo for __CreateParams {
        const ROUTE: &'static [u8] = &[24u8, 67u8, 114u8, 101u8, 97u8, 116u8, 101u8];
        type Params = Self;
        const ASYNC: bool = false;
    }
    #[derive(sails_rs::TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum ConstructorsMeta {
        Create(__CreateParams),
    }
}
#[cfg(target_arch = "wasm32")]
pub mod wasm {
    use super::*;
    use sails_rs::{gstd, hex, prelude::*};
    static mut PROGRAM: Option<MyProgram> = None;
    #[unsafe(no_mangle)]
    extern "C" fn init() {
        use gstd::InvocationIo;
        let mut input: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
        let invocation_route = if let Ok(request) = meta_in_program::__CreateParams::decode_params(
            input,
        ) {
            let program = MyProgram::create();
            unsafe {
                PROGRAM = Some(program);
            }
            meta_in_program::__CreateParams::ROUTE
        } else {
            gstd::unknown_input_panic("Unexpected ctor", input)
        };
        gstd::msg::reply_bytes(invocation_route, 0).expect("Failed to send output");
    }
    #[unsafe(no_mangle)]
    extern "C" fn handle() {
        let mut input = gstd::msg::load_bytes().expect("Failed to read input");
        let program_ref = unsafe { PROGRAM.as_mut() }.expect("Program not initialized");
        if input.starts_with(&__ROUTE_SVC1) {
            let mut service = program_ref.service1();
            let is_async = service
                .check_asyncness(&input[__ROUTE_SVC1.len()..])
                .unwrap_or_else(|| {
                    gstd::unknown_input_panic(
                        "Unknown call",
                        &input[__ROUTE_SVC1.len()..],
                    )
                });
            if is_async {
                gstd::message_loop(async move {
                    service
                        .try_handle_async(
                            &input[__ROUTE_SVC1.len()..],
                            |encoded_result, value| {
                                gstd::msg::reply_bytes(encoded_result, value)
                                    .expect("Failed to send output");
                            },
                        )
                        .await
                        .unwrap_or_else(|| {
                            gstd::unknown_input_panic("Unknown request", &input)
                        });
                });
            } else {
                service
                    .try_handle(
                        &input[__ROUTE_SVC1.len()..],
                        |encoded_result, value| {
                            gstd::msg::reply_bytes(encoded_result, value)
                                .expect("Failed to send output");
                        },
                    )
                    .unwrap_or_else(|| gstd::unknown_input_panic(
                        "Unknown request",
                        &input,
                    ));
            }
        } else if input.starts_with(&__ROUTE_SERVICE2) {
            let mut service = program_ref.service2();
            let is_async = service
                .check_asyncness(&input[__ROUTE_SERVICE2.len()..])
                .unwrap_or_else(|| {
                    gstd::unknown_input_panic(
                        "Unknown call",
                        &input[__ROUTE_SERVICE2.len()..],
                    )
                });
            if is_async {
                gstd::message_loop(async move {
                    service
                        .try_handle_async(
                            &input[__ROUTE_SERVICE2.len()..],
                            |encoded_result, value| {
                                gstd::msg::reply_bytes(encoded_result, value)
                                    .expect("Failed to send output");
                            },
                        )
                        .await
                        .unwrap_or_else(|| {
                            gstd::unknown_input_panic("Unknown request", &input)
                        });
                });
            } else {
                service
                    .try_handle(
                        &input[__ROUTE_SERVICE2.len()..],
                        |encoded_result, value| {
                            gstd::msg::reply_bytes(encoded_result, value)
                                .expect("Failed to send output");
                        },
                    )
                    .unwrap_or_else(|| gstd::unknown_input_panic(
                        "Unknown request",
                        &input,
                    ));
            }
        } else {
            gstd::unknown_input_panic("Unexpected service", &input)
        };
    }
    #[unsafe(no_mangle)]
    extern "C" fn handle_reply() {
        use sails_rs::meta::ProgramMeta;
        if MyProgram::ASYNC {
            gstd::handle_reply_with_hook();
        }
    }
    #[unsafe(no_mangle)]
    extern "C" fn handle_signal() {
        use sails_rs::meta::ProgramMeta;
        if MyProgram::ASYNC {
            gstd::handle_signal();
        }
    }
}
